FixMath.h

```diff
@@ -0,0 +1,1047 @@
+    - all operations between fixed point number is safe (it won't overflow) and preserve the precision. In particular:
+    - only addition, subtraction and multiplication are implemented
+    - any operation between a signed and an unsigned leads to a signed number
+    - resulting numbers will be casted to a type big enough to store the expected values. It follows that it is worth starting with types that are as small as possible to hold the initial value.
+    - all operations between a fixed point number and a native type (int, float, uint) are *not* safe. If the resulting value cannot be represented in the fixed point type it will overflow. Only addition, subtraction, multiplication and right/left shift are implemented.
+    - safe right/left shifts, which return the correct value in the correct type are implemented as .sR<shift>() and .sL<shift>() respectively, shift being the shifting amount. These shifts are basically on
+
+    More specifically on the returned types of the operations between fixed point math types:
+    - Additions:
+    - UFixMath<NI,NF> + UFixMath<_NI,_NF> returns UFixMath<MAX(NI,_NI)+1,MAX(NF,_NF)>
+    - SFixMath<NI,NF> + SFixMath<_NI,_NF> returns SFixMath<MAX(NI,_NI)+1,MAX(NF,_NF)>
+    - UFixMath<NI,NF> + SFixMath<_NI,_NF> returns SFixMath<MAX(NI,_NI)+1,MAX(NF,_NF)>
+    - UFixMath<NI,NF> + anything_else returns UFixMath<NI,NF>
+    - SFixMath<NI,NF> + anything_else returns SFixMath<NI,NF>
+    - Subtractions:
+    - UFixMath<NI,NF> - UFixMath<_NI,_NF> returns SFixMath<MAX(NI,_NI),MAX(NF,_NF)>
+    - SFixMath<NI,NF> - SFixMath<_NI,_NF> returns SFixMath<MAX(NI,_NI)+1,MAX(NF,_NF)>
+    - SFixMath<NI,NF> - UFixMath<_NI,_NF> returns SFixMath<MAX(NI,_NI)+1,MAX(NF,_NF)>
+    - UFixMath<NI,NF> - anything_else returns UFixMath<NI,NF>
+    - SFixMath<NI,NF> - anything_else returns SFixMath<NI,NF>
+    - (-)SFixMath<NI,NF> return SFixMath<NI,NF>
+    - (-)UFixMath<NI,NF> return SFixMath<NI,NF>
+    - Multiplications:
+    - UFixMath<NI,NF> * UFixMath<_NI,_NF> returns UFixMath<NI+_NI,NF+_NF>
+    - UFixMath<NI,NF> * SFixMath<_NI,_NF> returns SFixMath<NI+_NI,NF+_NF>
+    - SFixMath<NI,NF> * SFixMath<_NI,_NF> returns SFixMath<NI+_NI,NF+_NF>
+    - UFixMath<NI,NF> * anything_else returns UFixMath<NI,NF>
+    - SFixMath<NI,NF> * anything_else returns SFixMath<NI,NF>
+    - Shifts:
+    - UFixMath<NI,NF> .sR<NS> returns UFixMath<NI-NS,NF+NS>
+    - UFixMath<NI,NF> .sL<NS> returns UFixMath<NI+NS,NF-NS>
+    - same for SFixMath.
+      
+   
+*/
+
+
+
+
+#ifndef FIXMATH2_H_
+#define FIXMATH2_H_
+
+#include "IntegerType.h"
+
+#define MOZZI_SHIFTR(x,bits) (bits > 0 ? (x >> (bits)) : (x << (-bits))) // shift right for positive shift numbers, and left for negative ones.
+#define MAX(N1,N2) (N1 > N2 ? N1 : N2)
```

Comment:
This may have a high chance of clashing with defines set in other libs. Either rename to MOZZI_MAX, or test ```#ifndef``` MAX before defining. While at it, I also suggest parentheses around N1 and N2 (these are not "passed" as values but rather as mere string tokens, here, so something like ```MAX(a+1, 2)``` will not work as expected.

Reply:
Will add the parentheses soon, thanks!